### PR TITLE
Support any dynamic credentials in VaultCredentialsProvider

### PR DESCRIPTION
**Pull request by kdubb**
This change is in preparation of support for dynamic Vault credentials for RabbitMQ. It renames the "database" dynamic credentials supports to use "dynamic" in the names in an effort to be clear about its use.

Vault dynamic credentials all have basicallly identical responses so the credentials provider should support any of them easily.

This does essentially 4 things.
1. Renames `VaultDatabaseCredentials`, `VaultDBManager` and supporting classes to use the naming  "Dynamic"
2. Renames & deprecates `databaseCredentialsRole` to `credentialsRole`
3. Adds `credentialsMount` config and adds this as a parameter to `VaultDynamicCredentialsManager.getDynamicCredentials`
4. Adds an `expires-at` property to the returned credentials which is calculated from the Vault lease duration.

### CODE REVIEW

**Comment by kdubb** [comment]
A follow on PR for the actual RabbitMQ changes can be viewed at https://github.com/kdubb/quarkus/commit/12715d13950196af64d9493ad3a57753e576c3ad.

It requires a change to the `smallrye-reactive-messaging-rabbitmq` connector that is currently awaiting a release.

**Comment by kdubb** [comment]
@vsevel Feedback on this would be great. Just wanting to know if this is a good direction that is acceptable to the team.

**Comment by vsevel** [comment]
@sberyozkin we have had a long chat with @kdubb yesterday about this PR. we have agreed on the direction.
@kdubb I won't be able to look at it again before next monday.

**Comment by kdubb** [comment]
@vsevel Can we get this reviewed and merged? I've rebased and updated to relevant docs.

The SmallRye RabbitMQ connector version I need to finish dynamic credentials support is now available. Just need to get this in and I can submit that PR.

**Comment by sberyozkin** [comment]
@kdubb 

I'm not sure there is a strong enough reason to introduce a breaking change with the `Renames VaultDatabaseCredentials, VaultDBManager and supporting classes to use the naming "Dynamic"`. Doc clarifications should be enough IMHO

File: docs/src/main/asciidoc/vault-datasource.adoc
Comment (sberyozkin) [inline docs/src/main/asciidoc/vault-datasource.adoc:281]:
```suggestion
Now that Vault knows how to create users in PostgreSQL, we just need to change the `mydatabase` credentials
```

File: extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/CredentialsProviderConfig.java
Comment (sberyozkin) [inline extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/CredentialsProviderConfig.java:16]: I haven't been able to find where it is enforced, please add a check

Comment (kdubb) [reply extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/CredentialsProviderConfig.java:16]: `VaultCredentialsProvider` line 36 ish starts if statements that configure based on the first defined value.

**Comment by sberyozkin** [comment]
@kdubb Sure if @vsevel supports it is fine

I haven't been able to find `VaultDynamicCredentialsType` - perhaps it is no longer necessary ?

**Comment by kdubb** [comment]
@sberyozkin `VaultCredentialsType` was for `credentials-types`. After our chat, we decided to remove and replace it with a simple string. Using a string allows it to support any of Vault's dynamic credentials providers as well as one's with custom mount paths.

We also agreed to add the `credentials-request-path` option which allows configuration to support literally any current or future Vault secret engines. Currently all of the Vault secret engines for dynamic credentials use the value `creds`; which is the default for this configuration value.

File: docs/src/main/asciidoc/vault-datasource.adoc
Comment (kdubb) [reply docs/src/main/asciidoc/vault-datasource.adoc:281]: Fixed!

**Comment by sberyozkin** [comment]
@kdubb Thanks for the explanation

**Review by sberyozkin — APPROVED** [review]
Please update the 2.6 migration guide

**Comment by gsmet** [comment]
@sberyozkin it's a wiki so @kdubb can't. @kdubb, if you provide a section in Markdown, I will add it there: https://github.com/quarkusio/quarkus/wiki/Migration-Guide-2.6 .

@sberyozkin @vsevel if this one is ready, I would prefer we merge it really soon as I'd like it to be in before I move the Vault extension outside of the Core. I plan to do that either tomorrow afternoon or on Monday.

**Comment by vsevel** [comment]
hello @kdubb. thanks for this contribution. I will go through our chat (there may be some elements from the discussion we can share here for transparency and would help explain the rationale plus support discussion) again tomorrow and spend some time on the review.

**Comment by kdubb** [comment]
@gsmet The only changes worth noting in migration are the deprecation of the `database-credentials-type` configuration value; as it is only deprecated, the original value still works.

Suggested excerpt for the migration guide:


## Datasource Credentials Provider
The configuration for dynamic datasource credentials was changed as more more types of dynamic credentials are supported. Specifically the `database-credentials-role` key for configuring a credentials provider was deprecated. It has been replaced with two keys `credentials-role` and `credentials-type`.

Using the example from the [Dynamic Database Credentials](https://quarkus.io/guides/vault-datasource#dynamic-database-credentials) guide. The previously configuration of:
```properties
quarkus.vault.credentials-provider.mydatabase.database-credentials-role=mydbrole
```

Changes to:
```properties
quarkus.vault.credentials-provider.mydatabase.credentials-role=mydbrole
quarkus.vault.credentials-provider.mydatabase.credentials-type=database
```

Note that the previous configuration key of `database-credentials-role` has only been deprecated and will continue to work.

**Review by vsevel — CHANGES_REQUESTED** [review]
Hello @kdubb I have mainly 2 comments:
- I think we should have a credentials type config parameter separated from the mount point
- Then we should probably have calculated defaults for the mount and forced values for the request path based on this credentials type

open to discussion.

File: extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultCredentialsProvider.java
Comment (kdubb) [reply extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultCredentialsProvider.java:42]: Currently all of Vault's secret plugins dynamic credentials use `creds`; including Database and RabbitMQ. This is mainly for somebody who invents their own plugin or for some future Vault plugin that doesn't follow current conventions. Also, it does default to `creds` to make it optional; which is why I've left it out of all documentation. In the follow on PR for RabbitMQ, you only need to same two options of `credentials-role` and `credentials-type`.

File: extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/CredentialsProviderConfig.java
Comment (kdubb) [reply extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/CredentialsProviderConfig.java:46]: Currently, the way Vault works, you don't need a "type" because all credentials are returned in the exact same format; as it stands I don't see this changing. We could rename this to `credentialsMountPoint` but in Quarkus, and even in Vault vernacular, "type" is used to refer to the plugin via its default mount point, which is the name of the plugin. Additionally, all of the current secrets plugins for credentials support many roles per plugin which negates any real need mount the plugin multiple times in Vault. So `credentialsType` seemed to "fit" to me. In the end, I could really go either way.

**Comment by kdubb** [comment]
@vsevel Using `credentials-type` (or if renamed, `credentials-mount-point`), `credentials-role` and `credentials-request-path` is less pedantic but it supports any conceivable secrets plugin without additional code. This includes multiple versions of the same plugin mounted at custom paths or even custom plugins.

If we were to be more pedantic and have a `credentials-type` with selectable `credentials-mount-point` and calculate it's request path then supporting a plugin down the road becomes a code change. That's what I'm trying to avoid with this style. I don't want to have to change Quarkus to support each dynamic credentials secrets plugin. As I said above, currently `credentials-request-path` is always `creds` and might always be `creds`; it's only there to future proof the code in the event some new plugin shows up that doesn't use it.

So, in my opinion, the only thing I think we should be considering is a name change of the `credentials-type` option to `credentials-mount-path`.

**Comment by kdubb** [comment]
Actually, if we were to change `credentials-type`, I would probably vote for the shorter `credentials-mount`.

**Review by gsmet — DISMISSED** [review]
Please coordinate with me (@gsmet) before merging this as we don't want to merge this after I moved the code to the Quarkiverse.

File: extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultDynamicCredentialsManager.java
Comment (kdubb) [reply extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultDynamicCredentialsManager.java:120]: A naming overlap between the public types and the internal dto types.

File: extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultDynamicCredentialsManager.java
Comment (kdubb) [reply extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultDynamicCredentialsManager.java]: fixed

File: integration-tests/vault/src/test/java/io/quarkus/vault/VaultITCase.java
Comment (kdubb) [reply integration-tests/vault/src/test/java/io/quarkus/vault/VaultITCase.java]: fixed

File: extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultCredentialsProvider.java
Comment (vsevel) [reply extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultCredentialsProvider.java:42]: ok. that was not obvious to me `creds` was used in the other impl as well. so I agree we can leave it as a default.

**Comment by vsevel** [comment]
@kdubb I vote for replacing the type by `credentials-mount`. it is closer to the real semantic, and makes sense when used along the requestn path attribute. but as you said, usually in vault the mount path is the type of the engine. I suppose circumstances are rare when you need to use another mount path (an example is having multiple kv engines with different versions and/or config). 
anyway, I could go either way as well. I will leave you with the final word on this.
please do not forget to squash. thanks.

**Review by vsevel — APPROVED** [review]

**Comment by kdubb** [comment]
@vsevel Renamed `credentials-type` to `credentials-mount`. Ultimately the decision rested with the fact that all the internal code was already referring to it as "mount". So now its uniform and it was ultimately a small change that only affected config key.

**Comment by kdubb** [comment]
@gsmet Squished and rebased and I updated to migration blurb above.

**Comment by gsmet** [comment]
Let's merge this, thanks!

**Pull request merged**
